dmriprep/workflows/base.py

```diff
@@ -331,176 +437,620 @@ def init_single_subject_wf(
+        LOGGER.info('\n'.join(rationale))
+        return groups
+
+    # group by warped space, defined by pe_dir and fmap
+    groups = defaultdict(list)
+    for ref_file, pe_dir, fmap, fmap_type in parsed_dwis:
+        groups[(pe_dir, fmap)].append(ref_file)
+
+    # Also group by PE dir
+    pedir_groups = defaultdict(list)
+    for warp_group, member_images in groups.items():
+        pe_dir, _ = warp_group
+        pedir_groups[pe_dir].extend(member_images)
+
+    if ignore_fieldmaps:
+        output_groups = []
+        rationale.append("\nFieldmaps were ignored and images were grouped by PE Direction")
+        for pe_dir, member_images in pedir_groups.items():
+            output_groups.append({
+                'dwi_series': member_images,
+                'fieldmap_info': {'suffix': None},
+                'dwi_series_pedir': pe_dir})
+        LOGGER.info('\n'.join(rationale))
+        return output_groups
+
+    # Print the group memberships
+    rationale.append("\nThe following warped spaces were defined to contain:")
+    for warp_group, member_images in groups.items():
+        rationale.append(' * PEDir: %s, fmap: %s ' % warp_group)
+        for member_image in member_images:
+            rationale.append('    -> ' + member_image)
+
+    # group by pe dir and fmap
+    complete_groups = []
+    for warp_group, member_images in groups.items():
+        pe_dir, fmap_file = warp_group
+        complete_groups.append({
+            'dwi_series': member_images,
+            'fieldmap_info': fmap_pointers[fmap_file],
+            'dwi_series_pedir': pe_dir})
+
+    if prefer_dedicated_fmaps:
+        rationale.append('\nStrictly using data from fmaps/ for SDC')
+        LOGGER.info('\n'.join(rationale))
+        return complete_groups
+
+    # Ignoring dedicated fieldmaps, define warped spaces by PEDir
+    match_header = 'No opportunities for bidirectional PEPolar DWI series were found.'
+    group_keys = set(list(pedir_groups.keys()))
+    final_groups = []
+    matches_rationale = []
+    while len(group_keys):
+        key = group_keys.pop()
+        scan_list = pedir_groups[key]
+        pe_axis = key[0]
+
+        # Find opposing fieldmap groups
+        matches = [k for k in group_keys if k[0] == pe_axis]
+        num_matches = len(matches)
+
+        if num_matches == 1:
+            match_header = '\nThe following warp groups were used to SDC each other'
+            # Found a perfect match
+            match = matches[0]
+            match_scan_list = pedir_groups[match]
+            if using_eddy:
+                matches_rationale.append(
+                    '  * A single group containing %s/%s PE dir groups was sent to TOPUP' % (
+                        key, match))
+                final_groups.append(
+                    {'dwi_series': scan_list,
+                     'fieldmap_info': {'suffix': 'rpe_series', 'rpe_series': match_scan_list,},
+                     'dwi_series_pedir': pe_dir})
```

Comment:
This should be handled by the downstream workflow. I'm sure that, given the pe dir groups split the workflow could combine them for TOPUP.